src/mui/input/EmbeddedArrayInput.js

```diff
@@ -0,0 +1,161 @@
+ *
+ *
+ * Use it with any set of input componentents as children, like `<TextInput>`,
+ * `<SelectInput>`, `<RadioButtonGroupInput>` ... etc.
+ *
+ * @example
+ * export const CommentEdit = (props) => (
+ *     <Edit {...props}>
+ *         <SimpleForm>
+ *              <EmbeddedArrayInput source="links">
+ *                  <SimpleList>
+ *                       <TextField source="url" />
+ *                       <TextField source="context"/>
+ *                   </SimpleList>
+ *               </EmbeddedArrayInput>
+ *         </SimpleForm>
+ *     </Edit>
+ * );
+ */
+export class EmbeddedArrayInput extends Component {
+    constructor(props) {
+        super(props);
+        // nothing to do for now
+    }
+
+    renderList = ({ fields }) => {
+        const { children, arrayElStyle } = this.props;
+        return (
+            <div className="EmbeddedArrayInputContainer" style={{ margin: '1em' }}>
+                <div>
+                    {
+                        fields.map((member, index) =>
+                            <ArrayElement key={index} fields={fields} inputs={children} elStyle={arrayElStyle} member={member} index={index} />
+                        )
+                    }
+                </div>
+                <FlatButton primary style={{ float: 'right' }} icon={<ContentCreateIcon />} label="Add" onClick={() => fields.push({})} />
+                <div style={{ clear: 'both' }} />
+            </div>
+        )
+    }
+
+
+    render() {
+        const { input, resource, label, source, allowEmpty, basePath, onChange, children, meta } = this.props;
+
+        return (
+            <FieldArray name={source} component={this.renderList} />
+        )
+    }
+}
+
+const ArrayElement = ({ fields, inputs, member, index, elStyle }) => {
+    const removeElement = () => fields.remove(index);
+    const containerStyle = { padding: '0 1em 1em 1em' };
+    return (
+        <div style={elStyle} className="EmbeddedArrayInputItemContainer">
+            <div style={containerStyle}>
+                {
+                    React.Children.map(inputs, input => input && (
+                        <div key={input.props.source} className={`aor-input-${input.props.source}`} style={input.props.style}>
+                            <ArrayElementFormField input={input} prefix={member} />
+                        </div>
+                    ))
+                }
+            </div>
+            <FlatButton
+                primary
+                label="Remove"
+                icon={<ActionDeleteIcon />}
+                onClick={removeElement} />
+        </div>
+    )
+}
+
+import Labeled from './Labeled';
+import { required } from '../form/validate';
+
+const isRequired = (validate) => {
+    if (validate === required) return true;
+    if (Array.isArray(validate)) {
+        return validate.includes(required);
+    }
+    return false;
+};
+
+const ArrayElementFormField = ({ input, prefix, ...rest }) => {
```

Comment:
Disregard this one, you've already answered in the PR description. But adding your explanation as a comment in the code would help remember the reason.